docs: add new documentation related to testing style guide and testing template

EshaanAgg opened the pull request:
## Summary
Adds 
- Style guide for testing
- A test template and common ways to use the mocks

## References
[Slack Discussion](https://learningequality.slack.com/archives/C03S2EN192A/p1712478257893569)

## Reviewer guidance


----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

EshaanAgg committed:
add frontend testing style guide

EshaanAgg committed:
add files

EshaanAgg committed:
add testing layout

EshaanAgg commented on docs/frontend_architecture/unit_testing.rst:
```diff
@@ -21,4 +21,79 @@ Tests are written in JavaScript, and placed in the 'assets/test' folder. An exam
+If you have never worked with Vue Testing Library's (VTL) before, it is highly recommeed to the `VTL examples page <https://testing-library.com/docs/vue-testing-library/examples>`__ to see the library in action and understand it's use. If you're comfortable with the same, you can either:
+
+-  Read some of the exisiting tests in Kolibri to see how we make use of the same.
+-  Checkout the `test template <TODO>`__ and the `style guide <TODO>`__ to start writing your own!
```
I am unsure as to how to add an internal link and external link to documentation on an `.rst` file. Would appreciate any help with the same.

AlexVelezLl replied:
I think we can use relative paths e.g. https://github.com/EshaanAgg/kolibri/blob/0748571b4a6f049fb305ed5a712891baf8011640/docs/development_workflow.rst#L117

EshaanAgg replied:
Thanks!

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

EshaanAgg committed:
fix links

EshaanAgg committed:
Merge branch 'doc' of https://github.com/EshaanAgg/kolibri into doc

AlexVelezLl commented on docs/frontend_architecture/unit_testing.rst:
```diff
@@ -21,4 +21,79 @@ Tests are written in JavaScript, and placed in the 'assets/test' folder. An exam
 
 
-Vue.js components can also be tested. The management plugin contains an example (*kolibri/plugins/management/assets/test/management.js*) where the component is bound to a temporary DOM node, changes are made to the state, and assertions are made about the new component structure.
+For the frontend testing of ``Vue.js`` components, we make use of `Vue Testing Library <https://testing-library.com/docs/vue-testing-library/intro/>`__
```
In the example above, I think we can update the test example with one more updated and closer to what we normally do.

EshaanAgg replied:
Alright!

AlexVelezLl commented on docs/frontend_architecture/unit_testing.rst:
```diff
@@ -21,4 +21,79 @@ Tests are written in JavaScript, and placed in the 'assets/test' folder. An exam
 
-Vue.js components can also be tested. The management plugin contains an example (*kolibri/plugins/management/assets/test/management.js*) where the component is bound to a temporary DOM node, changes are made to the state, and assertions are made about the new component structure.
+For the frontend testing of ``Vue.js`` components, we make use of `Vue Testing Library <https://testing-library.com/docs/vue-testing-library/intro/>`__
+and the associated ecosystem.
```
I would change this introduction sentence to something like:
`We use Vue Testing Library (VTL) and its associated ecosystem to test Vue components, as it allows us to test UI components in a user-centric way.` (And define the "VTL" term in the first place this appears in the doc.)

EshaanAgg replied:
Got it! Thanks for the review. Will update.

AlexVelezLl commented on docs/frontend_architecture/unit_testing.rst:
```diff
@@ -21,4 +21,79 @@ Tests are written in JavaScript, and placed in the 'assets/test' folder. An exam
+For the frontend testing of ``Vue.js`` components, we make use of `Vue Testing Library <https://testing-library.com/docs/vue-testing-library/intro/>`__
+and the associated ecosystem.
+
+It is based on the philosphy that “The more your tests resemble the way your software is used, the more confidence they can give you." Rather than dealing with instances of rendered Vue components, it allows our tests to work with actual DOM nodes in the same way the user would. This deals to more user centric and a better quality of tests generally, and thus we have been making efforts to use th ...
```
We have a typo here:  `philosophy`.

We can also rewrite the second sentence to be something like `Rather than dealing with instances of rendered Vue components, it allows our tests to work with actual DOM nodes and simulate interactions the same way the user would.`. And I think we can remove the third sentence. I think in the last sentence of the section we could add a heads up that in the past we used Vue test utils, but we are changing it in favor or Vue testing library.

EshaanAgg replied:
Got it!

AlexVelezLl commented on docs/frontend_architecture/unit_testing.rst:
```diff
@@ -21,4 +21,79 @@ Tests are written in JavaScript, and placed in the 'assets/test' folder. An exam
+
+It is based on the philosphy that “The more your tests resemble the way your software is used, the more confidence they can give you." Rather than dealing with instances of rendered Vue components, it allows our tests to work with actual DOM nodes in the same way the user would. This deals to more user centric and a better quality of tests generally, and thus we have been making efforts to use th ...
+
+If you have never worked with Vue Testing Library's (VTL) before, it is highly recommeed to the `VTL examples page <https://testing-library.com/docs/vue-testing-library/examples>`__ to see the library in action and understand it's use. If you're comfortable with the same, you can either:
```
We can be more direct here I think. And say something like `To learn more about VTL, you can check its examples page. You can also check out our testing templates and our style guide to start writing new test suites in Kolibri.`

AlexVelezLl commented on docs/frontend_architecture/unit_testing.rst:
```diff
@@ -21,4 +21,79 @@ Tests are written in JavaScript, and placed in the 'assets/test' folder. An exam
+Naming Conventions
+~~~~~~~~~~~~~~~~~~
+
+Each folder is expected to have a ``__tests__`` folder, which would contain all the test files for the files in the same (whether ``.js`` or ``.vue``). The test files are expected to follow the naming convention ``<Name of the file being tested>.spec.js``.
```
Here we need to define better this, as in the beginning of the page we encourage to use the `assets/test` folder, but it seems we have been mixing it. cc: @rtibbles @MisRob

rtibbles replied:
The separate `tests` folder was a convention that I created, and has been largely superseded by the more 'Jest-y' `__tests__` folder, so I would be fine with this being the overall guidance, but yes, it should be consistently given!

EshaanAgg replied:
Got it. So should I work on removing the `tests` folder recommendation?

MisRob replied:
> So should I work on removing the tests folder recommendation

Yes, you can replace it by `__tests__` located within a close proximity to a component/utility being tested

AlexVelezLl commented on docs/frontend_architecture/unit_testing.rst:
```diff
@@ -21,4 +21,79 @@ Tests are written in JavaScript, and placed in the 'assets/test' folder. An exam
+Use of ``renderComponent`` function
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+All the files testing ``Vue.js`` components are expected to have their first function as ``renderComponent``. The job of this function should be to setup all the necesary mocks to render the component in the testing environment properly. It can also revieve an optional argument ``props``, which can be used to overwrite:
```
I think we can be more direct here: ```To avoid repeating boilerplate code while testing Vue components, define a `renderComponent` function to set up all the necessary mocks, stubs, and default props values to render the component to test.```

AlexVelezLl commented on docs/frontend_architecture/unit_testing.rst:
```diff
@@ -21,4 +21,79 @@ Tests are written in JavaScript, and placed in the 'assets/test' folder. An exam
+Queries
+~~~~~~~
+
+VTL provides a number of `queries <https://testing-library.com/docs/vue-testing-library/cheatsheet#queries>`__ that can be used to query the DOM nodes. There are primarily three types of queries: ``get``, ``query`` and ``find``. All three of them have two variants: one to query just a single DOM node, and the other to query multiple. Please ensure that you are using the correct kind of the query  ...
```
Not sure about rewriting this in our docs, I think we can refer to the testing library docs and say that we follow their recommended priority. What are your thoughts on this? @MisRob

MisRob replied:
I wouldn't mind having it briefly mentioned as a reminder since it's important, so perhaps something along the lines "Pay attention to the recommended priority of using queries: `get`, `query`, and `find`. See (link) for more." would balance both perspectives?

EshaanAgg replied:
Got it!

AlexVelezLl commented on docs/frontend_architecture/index.rst:
```diff
@@ -13,3 +13,4 @@ Frontend architecture
   dependencies
   unit_testing
   frontend_build_pipeline
```
I think we can have `unit_testing` and `testing_layout` together in the tree.

AlexVelezLl commented on docs/frontend_architecture/testing_layout.rst:
```diff
@@ -0,0 +1,176 @@
+1. You would need to replace the ``YourExampleComponent`` with the actual component you want to test.
+2. It is usually a good practice to have the first test as a smoke test in all your components.
+3. You can add describe blocks to add context to your tests. Remember to keep them as short as possible, and a test block should be responsible for testing only one functionality.
+4. It is advisable to define the default props as global constants so they are easily accessible in the tests (though this may vary from file to file).
```
I think we can remove this one.

EshaanAgg replied:
Okay!

AlexVelezLl reviewed (commented):
Thanks for all your work here @EshaanAgg! This is looking great, these are some thoughts from an initial reading. But it's definitely looking good.

EshaanAgg committed:
update docs

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

EshaanAgg commented:
Thanks, @MisRob and @AlexVelezLl, for all the help. I have updated the documentation as per the review comments. I also took the liberty to make two more changes:
- Remove the imports for VueRouter from each of the sample tests and templates, as the same should be fixed by #12105
- As discussed in [a review comment](https://github.com/learningequality/kolibri/pull/12056#discussion_r1570830460) of #12056, it is better that while mocking composables we pass the composable related mocking data in a separate property to the render function. I have updated that recommendation too. 

Let me know if you would like to see something else added!

EshaanAgg committed:
update links in docs

EshaanAgg committed:
update links in docs

EshaanAgg committed:
fix link finally

AlexVelezLl commented on docs/frontend_architecture/unit_testing.rst:
```diff
@@ -20,5 +19,77 @@ Tests are written in JavaScript, and placed in the 'assets/test' folder. An exam
+VTL provides a number of `queries <https://testing-library.com/docs/vue-testing-library/cheatsheet#queries>`__ that can be used to query the DOM nodes. There are primarily three types of queries: ``get``, ``query`` and ``find``. All of these queries have different variants, which are used to query the DOM nodes based on different criteria. Some examples of the same include: ``getByText``, ``query ...
+
+Making use of ``screen``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
I think we could change this to something like:

```rst
Making use of VTL ``screen`` object
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For querying DOM nodes, the preferred way is to use the ``screen`` object provided by ``@testing-library/vue``. So instead of destructuring the queries functions from the object returned by the VTL ``render`` function, you can use the VTL ``screen`` object for faster access to all its queries:

.. code-block:: javascript
// ❌
const {getByRole} = render(Example)
const errorMessage = getByRole('alert')

// ✅
render(<Example />)
const errorMessage = screen.getByRole('alert')
```

I think we can write some examples like these here to make it clearer what we want to communicate.

EshaanAgg replied:
Got it! This is a pretty neat way to communicate the same. Thanks!

AlexVelezLl commented on docs/frontend_architecture/unit_testing.rst:
```diff
@@ -20,5 +19,77 @@ Tests are written in JavaScript, and placed in the 'assets/test' folder. An exam
+
+Instead of importing the query functions from the object returned by the ``render`` function, you should import the ``screen`` object for the same. The ``screen`` object is a global object that is provided by the ``@testing-library/vue`` package, and it contains all the query functions as properties. This ensures that we do not have to import the query functions in test, and can call them directl ...
+
+``userEvent`` over ``fireEvent``
```
I would change this to: 

Prefer the ``userEvent`` package over ``fireEvent`` to simulate user interactions.

AlexVelezLl commented on docs/frontend_architecture/unit_testing.rst:
```diff
@@ -20,5 +19,77 @@ Tests are written in JavaScript, and placed in the 'assets/test' folder. An exam
+
+-  The default props passed to the component being rendered
+-  Configuration passed to other mocks/stubs (like the values the getters for store mock should return, arguments to Vue Router etc) according to the test case
+
```
Here we can add a simple example of this, to make it clearer just in case.

EshaanAgg replied:
Done!

AlexVelezLl commented on docs/frontend_architecture/unit_testing.rst:
```diff
@@ -20,5 +19,77 @@ Tests are written in JavaScript, and placed in the 'assets/test' folder. An exam
+Using ``testing-library/jest-dom``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+`testing-library/jest-dom <https://github.com/testing-library/jest-dom>`__ provides a set of custom jest matchers that extend jest to common usecases for frontend testing, like checking classes, attributes text content, CSS properties etc. The use of these matchers helps to make the tests more declarative and clear to read and maintain. Please make sure you use the appropiate matcher from the sam ...
```
We could add a couple of examples here about the `testing-library/jest-dom` jest matchers.

AlexVelezLl commented on docs/frontend_architecture/testing_layout.rst:
```diff
@@ -0,0 +1,186 @@
+Some quick notes about the code snippet:
+
+1. You would need to replace the ``YourExampleComponent`` with the actual component you want to test.
+2. It is usually a good practice to have the first test as a smoke test in all your components.
```
I think this list is actually more Style Guides, so I think is better to have these items as subItems of the Style Guides section:
1. Add smoke tests: Add a smoke test to every test suites that only renders the most basic example of a component, where the only thing about assert is that the simplest render does not throw an error.
2. Use describe blocks: Use describe blocks to group unit tests that test similar functionality.
3. Avoid long and complex unit tests: A unit test should be kept simple and test a single execution flow. So that it is easy for someone else to read the test and understand the functionality of the component.
4. Use default props that are not relevant to your unit test: Within your `renderComponents` function declare default props so that you do not have to define them in unit tests where they are not relevant. (Here we could also give an example about this, although it is less important than the others).
5. Explicitly declare props that are relevant to your unit test: Avoid using default props for inputs that are relevant to your unit tests, and instead declare them explicitly in your test. For example, if we are testing the rendering of a list, then let's explicitly declare the list in the unit test. Let's not wait until someone else reading the test has to go to the `renderComponent` function to see what the unit test input was.

The wording could change, but I've more or less written down the ideas that could go in these subsections. With this I would think that this list of "quick notes about the code snippet" could be eliminated. Let me know what you think.

EshaanAgg replied:
Yes, it makes complete sense. Reading these points explicitly under the style guide would have better effect in terms of communication of these ideas. Thanks for taking out the time to articulate all these points! Will work on updating them in the documentation.

AlexVelezLl commented on docs/frontend_architecture/testing_layout.rst:
```diff
@@ -0,0 +1,186 @@
+
+
+   // Helper function to render the component with Vuex store
+   const renderComponent = props => {
```
We can do something similar to the composables mocks here, to make it clearer what are store values, and what component props. So we could change a little bit this to be something like:

```javascript
const renderComponent = props => {
     const { store = {}, ...componentProps } = props;
     return render(TotalPoints, {
       store: {
         getters: {
           totalPoints: () => store.totalPoints ?? 0,
           currentUserId: () => store.currentUserId ?? "user-01",
           isUserLoggedIn: () => store.isUserLoggedIn ?? true,
         },
         actions: {
           fetchPoints: () => store.points ?? 0,
         },
       },
       props
     });
   };
```

Also, we can delete the `routes: new VueRouter()` prop from these examples, as it is no longer required :tada:

EshaanAgg replied:
Got it! Thanks!

AlexVelezLl reviewed (changes requested):
Hi @EshaanAgg! Thank you for all your great work here. Sorry for the delay in the review. And Sorry for the back and forth. I think this will be the last before approval from my side. It's pretty polished already :tada:.

EshaanAgg committed:
address review comments 1

EshaanAgg committed:
address review comments 2

AlexVelezLl reviewed (approved):
Thanks for all your hard work here @EshaanAgg!! :tada:. In the future we could edit the documentation and cover some point that we might be missing, but for now it looks pretty good. The style guide shows the what and how quite well. The testing template looks quite good and complete for multiple cases, and the structure in general seems adequate and easy to follow!! LGTM! Thank you :rocket::rocket::rocket:.
